devito/symbolics/manipulation.py

```diff
@@ -319,12 +319,92 @@ def reuse_if_untouched(expr, args, evaluate=False):
         return expr.func(*args, evaluate=evaluate)
 
 
-def evalmin(a, b):
+def evalrel(func=min, input=None, assumptions=None):
     """
-    Simplify min(a, b) if possible.
+    The purpose of this function is two-fold: (i) to reduce the `input` candidates of a
+    for a MIN/MAX expression based on the given `assumptions` and (ii) return the nested
+    MIN/MAX expression of the reduced-size input.
+
+    Parameters
+    ----------
+    func : builtin function or method
+        min or max. Defines the operation to simplify. Defaults to `min`.
+    input : list
+        A list of the candidate symbols to be simplified. Defaults to None.
+    assumptions : list
+        A list of assumptions formed as relationals between candidates, assumed to be
+        True. Defaults to None.
+
+    Examples
+    --------
+    Assuming no values are known for `a`, `b`, `c`, `d` but we know that `d<=a` and
+    `c>=b` we can safely drop `a` and `c` from the candidate list.
+
+    >>> from devito import Symbol
+    >>> a = Symbol('a')
+    >>> b = Symbol('b')
+    >>> c = Symbol('c')
+    >>> d = Symbol('d')
+    >>> evalrel(max, [a, b, c, d], [Le(d, a), Ge(c, b)])
+    MAX(a, c)
     """
+    sfunc = (Min if func is min else Max)  # Choose SymPy's Min/Max
+
+    # Form relationals so that RHS has more arguments than LHS:
+    # i.e. (a + d >= b) ---> (b <= a + d)
+    assumptions = [a if len(a.rhs.args) > len(a.lhs.args) else
+                   a.reversed for a in as_list(assumptions)]
+
+    # Expand relations if possible
+    processed = []
+    for a in as_list(assumptions):
+        if isinstance(a, (Ge, Gt)) and a.rhs.is_Add and a.lhs.is_positive:
+            if all(i.is_positive for i in a.rhs.args):
+                # If `c >= a + b, {a, b, c} >= 0` then add 'c>=a, c>=b'
+                processed.extend(Ge(a.lhs, i) for i in a.rhs.args)
+            elif len(a.rhs.args) == 2:
+                # If `c >= a + b, a>= 0, b<=0` then add 'c>=b, c<=a'
+                processed.extend(Lt(a.lhs, i) if i.is_positive else
+                                 Gt(a.lhs, i) for i in a.rhs.args)
+            else:
+                processed.append(a)
+        else:
+            processed.append(a)
+
+    # Apply assumptions to fill a subs mapper
+    mapper = {}
+    for a in processed:
+        if set(a.args).issubset(input):
```

Comment:
can you write an example of when we enter this `if`?

Reply:
Added.

Reply:
is this `if` really necessary? I have a feeling it should work regardless